presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -384,14 +446,68 @@ private RowExpression notCallExpression(RowExpression argument)
         {
             return new CallExpression("not", functionResolution.notFunction(), BOOLEAN, singletonList(argument));
         }
+    }
 
-        private boolean isConjunctionOrDisjunction(RowExpression expression)
+    private class ConvertNormalFormVisitor
+            implements RowExpressionVisitor<RowExpression, Form>
+    {
+        @Override
+        public RowExpression visitSpecialForm(SpecialFormExpression specialForm, Form clauseJoiner)
         {
-            if (expression instanceof SpecialFormExpression) {
-                Form form = ((SpecialFormExpression) expression).getForm();
-                return form == AND || form == OR;
+            if (!isConjunctionOrDisjunction(specialForm)) {
+                return specialForm;
+            }
+
+            // normalize arguments
+            RowExpression left = specialForm.getArguments().get(0).accept(new ConvertNormalFormVisitor(), clauseJoiner);
+            RowExpression right = specialForm.getArguments().get(1).accept(new ConvertNormalFormVisitor(), clauseJoiner);
+
+            // check if already in correct form
+            if (specialForm.getForm() == clauseJoiner) {
+                return clauseJoiner == AND ? and(left, right) : or(left, right);
             }
-            return false;
+
+            // else, we expand and distribute
```

Comment:
What about the following:
```
Add a comment like:
```
// else, we expand and rewrite based on distributive property of Boolean algebra, for example
// (l1 OR l2) AND (r1 OR r2) <=> (l1 AND r1) OR (l1 AND r2) OR (l2 AND r1) OR (l2 AND r2) 
```

```